core/src/main/java/com/scalar/db/storage/cosmos/CosmosTableMetadata.java

```diff
@@ -2,52 +2,58 @@
 
 import com.google.common.base.MoreObjects;
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import java.util.Collections;
 import java.util.LinkedHashSet;
-import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
-import javax.annotation.concurrent.NotThreadSafe;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
 
 /**
  * A metadata class for a table of ScalarDB to know the type of each column
  *
  * @author Yuji Ito
  */
 @SuppressFBWarnings({"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
-@NotThreadSafe
+@Immutable
 public class CosmosTableMetadata {
-  private String id;
-  private LinkedHashSet<String> partitionKeyNames;
-  private LinkedHashSet<String> clusteringKeyNames;
-  private Map<String, String> clusteringOrders;
-  private Set<String> secondaryIndexNames;
-  private Map<String, String> columns;
-
-  public CosmosTableMetadata() {}
-
-  public void setId(String id) {
-    this.id = id;
-  }
-
-  public void setPartitionKeyNames(List<String> partitionKeyNames) {
-    this.partitionKeyNames = new LinkedHashSet<>(partitionKeyNames);
-  }
-
-  public void setClusteringKeyNames(List<String> clusteringKeyNames) {
-    this.clusteringKeyNames = new LinkedHashSet<>(clusteringKeyNames);
-  }
-
-  public void setClusteringOrders(Map<String, String> clusteringOrders) {
-    this.clusteringOrders = clusteringOrders;
+  private final String id;
+  private final LinkedHashSet<String> partitionKeyNames;
+  private final LinkedHashSet<String> clusteringKeyNames;
+  private final Map<String, String> clusteringOrders;
+  private final Set<String> secondaryIndexNames;
+  private final Map<String, String> columns;
+
+  public CosmosTableMetadata() {
+    this(null, null, null, null, null, null);
```

Comment:
Jackson requires a default constructor so we need it. I will add a comment.

Reply:
I see. Where are we using Jackson databind for CosmosTableMetadata? I want to reproduce it and look into it as I'm a bit familiar with Jackson databind.

Reply:
Ooops. Sorry I missed the background.

> This PR makes the CosmosMetadata and Record objects immutable. These DTO (Data Transfer Object) are used by the Cosmos SDK which uses Jackson.

Reply:
JFYI: The following might be an option to work with Jackson databind internally used in Cosmos SDK, although it would make things a bit complicated.

- build.gradle
```
    implementation 'com.fasterxml.jackson.core:jackson-databind:2.15.2'
```

- CosmosTableMetadata.java
```java
  public CosmosTableMetadata(
      @JsonProperty("id")
      @Nullable String id,
      @JsonProperty("partitionKeyNames")
      @Nullable LinkedHashSet<String> partitionKeyNames,
      @JsonProperty("clusteringKeyNames")
      @Nullable LinkedHashSet<String> clusteringKeyNames,
      @JsonProperty("clusteringOrders")
      @Nullable Map<String, String> clusteringOrders,
      @JsonProperty("secondaryIndexNames")
      @Nullable Set<String> secondaryIndexNames,
      @JsonProperty("columns")
      @Nullable Map<String, String> columns) {
    this.id = id != null ? id : "";
    this.partitionKeyNames = partitionKeyNames != null ? partitionKeyNames : new LinkedHashSet<>();
    ...
  }
```

I'm wondering why Cosmos SDK doesn't use https://github.com/FasterXML/jackson-modules-java8/tree/master/parameter-names or equivalent on their side...

Reply:
I see. 
Yes, it would be nice if the SDK were using that feature of Jackson. 
Though, adding the jackson datababind dependency on our side to tinker with the Cosmos SDK default behavior seems a bit dangerous to me.